Implement the "logs" endpoint for the journald log driver

nalind opened the pull request:
This change set uses the journald libraries to read log entries from the journal, and hooks that up to the "logs" endpoint.  It moves handling of the --last option out of the endpoint logic and into the log drivers themselves, because doing so has the potential cut down on how much data the journald driver will have to dump to its pipe.

It's multiple patches to hopefully be easier to review, but I have no qualms about squashing things together if that's preferred.

rhatdan commented:
Any comment from upstream?  This would be a really nice feature.  I already have logging people asking me about it.

thaJeztah commented:
Think they've been busy bug-hunting for the release.

LK4D4 commented:
I get `Unrecognized input header` with this driver. Also I have double output for `json-file` log-driver on first run of `docker logs -f`.

nalind commented:
I can't reproduce the first symptom; can you give me more details about how you're getting it to happen?  The duplicate log data doesn't appear to be driver-specific (now that I know what to look for, I'm seeing it with the journald driver, too), but it goes away if I revert 153f98bad51e2f46cf3853d2da0173bcfb4b687d.

LK4D4 commented:
@nalind I run
```
docker run -d --log-driver=journald busybox ping google.com
```
and then just do `docker logs -f` on that container. Nothing in daemon logs also.

nalind commented:
Ah, thanks!  It's still not giving me the ``Unrecognized input header`` error, though it does show me the first few lines of the output twice if I run ``docker logs -f`` quickly enough.

It looks as though, if the client and daemon disagreed about whether or not the container had a tty allocated to it, that the data stream coming back from the server wouldn't be muxed using an stdcopy.StdWriter, so it wouldn't be in the format that stdcopy.StdCopy expects to demux, and that should cause the ``Unrecognized input header`` message to appear, but I'm at a loss as to how to actually trigger it.

LK4D4 commented:
@nalind I'm using systemd-220. Maybe it returns crap instead of logs sometimes.

nalind commented:
I'm also running with version 220 (Fedora 23's systemd-220-3.fc23), but the like messages we retrieve from json-logs, data we read back from the journal should be wrapped up and chunked before it gets to the client, so if the driver was getting garbage back, I'd expect to get garbage in the output of the ``docker logs`` command rather than an error parsing the HTTP stream that's carrying it.  Is there anything else that might be making a difference?

LK4D4 commented:
@nalind I can't imagine. I'll try to put some logging to your code later.

nalind commented:
Ah, found it.  When built without the ``journald`` tag, which the build scripts try to enable automatically when the libraries are detected, this is triggered when the daemon sends back an error message indicating that it couldn't read the logs, because it sends that message back without the framing that the client is expecting.

The Dockerfile installs ``libsystemd-journal-dev``, and ``hack/make.sh`` tries to check for it, but that check fails because ``pkg-config`` isn't also available, so the ``journald`` build tag wasn't being enabled.  Adding it as a requirement to the top-level ``Dockerfile`` should handle most of the cases that the changes to ``contrib/builder/deb/generate.sh`` and ``hack/make/.build-rpm/docker-engine.spec`` didn't already cover.

As for the formatting of the error when it's sent back to the client, if ``ContainerLogs`` replaces the ``OutStream`` member of the ``ContainerLogsConfig`` that it gets with its wrapper, and ``getContainersLogs`` sends the error message through that, the client displays the error that the server sent.

I'm tweaking the patch set to fix both of these problems and rebasing against the current master.

nalind commented:
Rebasing to resolve reported merge conflicts.

nalind commented:
Catching up with the new fluentd driver.

thaJeztah commented:
ping @LK4D4 have you been able to successfully test this?

cpuguy83 commented:
Hi, the log reader interface was refactored and should (I hope!) be much simpler to implement now, as all you need to do is implement a `ReadLogs` function on journald.

LK4D4 commented:
Yes, I waited for this refactoring. Now implementation of journald logs should be simpler and more robust.

nalind commented:
The current refactor keeps knowledge of the number of lines to read when it's in "last" mode, and the cutoff time in "since" mode, in `ContainerLogs`.  The `journald` driver could definitely use that information to skip sending data that is only going to be ignored by `ContainerLogs`, and I suspect other log readers would, too.  (The current `ContainerLogs` also spends some time dealing with file numbers, which really seems to me like a `jsonfilelog`-specific notion, but a reader can ignore that.)

cpuguy83 commented:
@nalind Not anymore. ContainerLogs does nothing except call the ReadLogs on the configured logger and Listens for logger.Messages on a channel.

nalind commented:
@cpuguy83 Oh, you're right, that's new today, and it's *much* better.

thaJeztah commented:
> Oh, you're right, that's new today, and it's much better.

glad you like it; kudos to @cpuguy83 :+1:

nalind committed:
Add log reading to the journald log driver

If a logdriver doesn't register a callback function to validate log
options, it won't be usable.  Fix the journald driver by adding a dummy
validator.

Teach the client and the daemon's "logs" logic that the server can also
supply "logs" data via the "journald" driver.  Update documentation and
tests that depend on error messages.

Add support for reading log data from the systemd journal to the
journald log driver.  The internal logic uses a goroutine to scan the
journal for matching entries after any specified cutoff time, formats
the messages from those entries as JSONLog messages, and stuffs the
results down a pipe whose reading end we hand back to the caller.

If we are missing any of the 'linux', 'cgo', or 'journald' build tags,
however, we don't implement a reader, so the 'logs' endpoint will still
return an error.

Make the necessary changes to the build setup to ensure that support for
reading container logs from the systemd journal is built.

Rename the Jmap member of the journald logdriver's struct to "vars" to
make it non-public, and to make it easier to tell that it's just there
to hold additional variable values that we want journald to record along
with log data that we're sending to it.

In the client, don't assume that we know which logdrivers the server
implements, and remove the check that looks at the server.  It's
redundant because the server already knows, and the check also makes
using older clients with newer servers (which may have new logdrivers in
them) unnecessarily hard.

When we try to "logs" and have to report that the container's logdriver
doesn't support reading, send the error message through the
might-be-a-multiplexer so that clients which are expecting multiplexed
data will be able to properly display the error, instead of tripping
over the data and printing a less helpful "Unrecognized input header"
error.

Signed-off-by: Nalin Dahyabhai <nalin@redhat.com> (github: nalind)

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "journald-logs" git@github.com:nalind/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

rhatdan commented:
Can we get some movement on this.  Would love to get this into docker-1.9.

@cpuguy83 @crosbymichael @jfrazelle @tianon @LK4D4

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "journald-logs" git@github.com:nalind/docker.git somewhere
$ cd somewhere
$ git rebase -i HEAD~5
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

cpuguy83 commented:
@LK4D4 Do you think this is good to move to code-review?
I think this makes sense.

cpuguy83 commented on api/client/logs.go:
```diff
@@ -37,8 +37,8 @@ func (cli *DockerCli) CmdLogs(args ...string) error {
 
-	if logType := c.HostConfig.LogConfig.Type; logType != "json-file" {
-		return fmt.Errorf("\"logs\" command is supported only for \"json-file\" logging driver (got: %s)", logType)
+	if logType := c.HostConfig.LogConfig.Type; logType != "json-file" && logType != "journald" {
```
Probably best to just remove this check (and no need to pull the container at all as it does above).
The API will error out if the logger is now a supported.
Kinda weird we are doing this extra API call when the API errors anyway.

nalind replied:
Hmm, it looks like we still need to read the Tty field of the container config, in order to know whether or not we need to demux the data that's coming back.  And in cases where we would need to demux the data, but the logdriver implementation doesn't support reading on the server, I think the error that we currently send back isn't muxed, and that'll confuse the client.  I think that's how we got the "Unrecognized input header" error that @LK4D4 was seeing when the server was built without the libraries.

nalind replied:
Okay, went ahead and made that change, and tweaked getContainersLogs() and ContainerLogs() to avoid the "Unrecognized input header" error.

cpuguy83 commented on daemon/logger/journald/journald.go:
```diff
@@ -15,13 +16,22 @@ import (
 
 type journald struct {
-	Jmap map[string]string
+	Jmap    map[string]string
```
Can you document this, and probably give it a better name?

nalind replied:
Sure thing.  It'll be "vars", since they're additional variables and values that we want to add to the journaled log message.

calavera commented on daemon/logger/journald/read.go:
```diff
@@ -0,0 +1,303 @@
+	return
+}
+
+func (s *journald) ReadLogs(config logger.ReadConfig) *logger.LogWatcher {
```
do we need to implement this method for the unsupported platforms doing nothing? like we do for the `Close` method in `read_unsupported`?

nalind replied:
ContainerLogs() checks if the logger implements the LogReader interface, which contains only the ReadLogs() method, and returns a logger.ErrReadLogsNotSupported error directly if it doesn't, so I don't think we're supposed to implement the method when reading isn't supported.  That said, we could implement it and close the Msg channel immediately, if we had to.

calavera commented on pkg/graphdb/conn_sqlite3.go:
```diff
@@ -1,3 +1,5 @@
+// +build cgo
```
why is this change part of the PR?

nalind replied:
Yeah, it could probably stand to be in a different PR.

nalind replied:
Moved to PR#16022.

nalind commented:
Rebasing to clear a merge conflict and drop the patch which is now PR#16022.

calavera commented:
LGTM.

@nalind can you squash your commits into a single one so we can track all this feature together?

nalind commented:
Will do.

jfrazelle commented on contrib/builder/deb/ubuntu-debootstrap-utopic/Dockerfile:
```diff
@@ -0,0 +1,14 @@
+#
+# THIS FILE IS AUTOGENERATED; SEE "contrib/builder/deb/generate.sh"!
```
this file should not be here, i think you need a rebase

jfrazelle commented:
1. small rebase problem, other than that feel free to ping me after to merge thanks

LK4D4 commented:
@nalind I don't see nor ID, neither name of container in `journalctl -f`(which is probably not your problem)`, but I also see:
```
Error running logs job: configured logging reader does not support reading
```
on `docker logs`

nalind commented:
@LK4D4 They should show up if you use one of the verbose output formats (-o json, -o export, or -o verbose).

@jfrazelle Thanks for spotting that.  Dropped the file, rebased, and gave it another once over.

LK4D4 commented:
@nalind okay, see it now. What about `docker logs`? :)

nalind commented:
@LK4D4 The logic's compiled in when we find the headers (make.sh sets a build tag if pkg-config finds them) and since I don't get static libraries with libsystemd-journal-dev or systemd-devel packages, it also limits itself to non-static builds.  Are you testing with a static binary?

LK4D4 commented:
@nalind no, I'm testing with dynamic binary. Just noticed that it requires buildtag, I'll try to build with it.

jfrazelle commented:
im testing the debs/rpm build now :)

LK4D4 commented:
Yup, it works with tag. And I can't find static systemd libs too.

jfrazelle commented:
so can we not compile a static binary with this? is there licensing problems?

LK4D4 commented:
```
Error running logs job: configured logging reader does not support reading
```
is confusing though, but I don't know what to do with it.
@jfrazelle Yes, there is no way to compile static binary with it as I see.

tianon commented on contrib/builder/deb/generate.sh:
```diff
@@ -55,6 +55,7 @@ for version in "${versions[@]}"; do
 		libapparmor-dev # for "sys/apparmor.h"
 		libdevmapper-dev # for "libdevmapper.h"
 		libsqlite3-dev # for "sqlite3.h"
+		libsystemd-journal-dev # for "sd-journal.h" and libraries
```
This package doesn't exist on `precise`, so it needs to be removed in the `if ... precise` block below.

jfrazelle replied:
also it is libsystemd-dev on stretch

tianon replied:
In that case, maybe we'd be better off with a `case` statement to add this one specifically?  ie, something like:
```bash
case "$suite" in
    precise) ;;
    stretch|sid) packages+=( libsystemd-dev ) ;;
    *) packages+=( libsystemd-journal-dev ) ;;
esac
```

jfrazelle replied:
+1

nalind replied:
@tianon  Using `case` for that makes sense to me, too.  Fixing.

jfrazelle commented on contrib/builder/deb/debian-stretch/Dockerfile:
```diff
@@ -4,7 +4,7 @@
 FROM debian:stretch
 
-RUN apt-get update && apt-get install -y bash-completion btrfs-tools build-essential curl ca-certificates debhelper dh-systemd git libapparmor-dev libdevmapper-dev libsqlite3-dev --no-install-recommends && rm -rf /var/lib/apt/lists/*
+RUN apt-get update && apt-get install -y bash-completion btrfs-tools build-essential curl ca-certificates debhelper dh-systemd git libapparmor-dev libdevmapper-dev libsqlite3-dev libsystemd-journal-dev --no-install-recommends && rm -rf /var/lib/apt/lists/*
```
this is  libsystemd-dev on stretch

nalind replied:
Okay, I think I've incorporated these.

tianon commented:
IMO, we're probably safe to disable this driver for static builds, especially since even if we can get libsystemd to link in statically, it likely won't work very well if we have host version mismatching (just like the libdevmapper issues we have with static binaries).

jfrazelle commented:
retesting building debs and rpms

jfrazelle commented:
ubuntu:wily is libsystemd-dev as well sorry

jfrazelle commented:
the rpms worked

nalind commented:
@jfrazelle  No problem.  Your build box is clearly faster than mine.  :)

jfrazelle commented:
LGTM

jfrazelle commented:
@LK4D4 good for you?

LK4D4 commented:
LGTM

razyr commented:
Just a question, but does this have any impact on or implications for the native GELF support that has been added with Docker 1.8?

noxiouz commented on daemon/logger/journald/read.go:
```diff
@@ -0,0 +1,303 @@
+	}
+	// Add a match to have the library do the searching for us.
+	cmatch = C.CString("CONTAINER_ID_FULL=" + s.vars["CONTAINER_ID_FULL"])
+	if cmatch == nil {
```
I'm sorry for commenting the merged PR, but how it could be nil?

nalind replied:
Yup, looks like C.CString() doesn't return a NULL pointer on out-of-memory conditions, so this block is unnecessary.

nalind replied:
Opened PR#16283 to drop the block.  Thanks!

philips commented:
We have an outstanding PR to the github.com/coreos/go-systemd package, which is used elsewhere in docker, to implement a journal API based on code Dan Mace tried to get working in 2014: https://github.com/coreos/go-systemd/pull/110

dmp1ce commented:
@LK4D4 

I'm not getting `Error running logs job: configured logging reader does not support reading` with docker version 1.9.1 on Ubuntu Vivid when I run `docker logs`. `journalctl` works fine.

Should this be fixed now? I'm not sure why I'm getting the error.

LK4D4 commented:
@dmp1ce Are you getting error or not? I'm not sure how your docker is built, but with `--log-driver=journalctl` all should work fine.

dmp1ce commented:
@LK4D4  Sorry, I AM getting the error. I updated my comment with the removed "not".

I'm installing using the instructions here: https://docs.docker.com/engine/installation/ubuntulinux/

I'm using the journald driver using Docker Compose and `docker run --log-driver=journald`. I'm not using `--log-driver=journalctl`.

LK4D4 commented:
@dmp1ce it should be fine if binary built as dynamic with journald build-tag. I think @jfrazelle or @tianon know more than me about which binary we ship for ubuntu.

bobrik commented:
I am using official docker-engine package for jessie and getting this message too:

```
Error running logs job: configured logging reader does not support reading
```

```
ivan@36com10:~$ ldd `which docker`
	linux-vdso.so.1 (0x00007ffdec1b1000)
	libapparmor.so.1 => /usr/lib/x86_64-linux-gnu/libapparmor.so.1 (0x00007fe08b90a000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007fe08b6ed000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007fe08b4e8000)
	libdevmapper.so.1.02.1 => /lib/x86_64-linux-gnu/libdevmapper.so.1.02.1 (0x00007fe08b29f000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007fe08aef6000)
	/lib64/ld-linux-x86-64.so.2 (0x0000562f5b55f000)
	libselinux.so.1 => /lib/x86_64-linux-gnu/libselinux.so.1 (0x00007fe08acd0000)
	libudev.so.1 => /lib/x86_64-linux-gnu/libudev.so.1 (0x00007fe08aac1000)
	libpcre.so.3 => /lib/x86_64-linux-gnu/libpcre.so.3 (0x00007fe08a853000)
	librt.so.1 => /lib/x86_64-linux-gnu/librt.so.1 (0x00007fe08a64a000)
```

Getting logs through `journalctl` works except for #15722.

LK4D4 commented:
@bobrik How do you start your container and what version do you use? Seems like works fine for me on master with
```
docker run -it --rm --log-driver=journald busybox ping google.com
```

tianon commented:
Yeah, as far as I know, logging to journald should be enabled in our Ubuntu
and Debian packages, since they're dynamic for sure.

We have libsystemd-journal-dev in
https://github.com/docker/docker/blob/0b7e9ef910b91a0a8cb334db852ad143c4660155/contrib/builder/deb/debian-jessie/Dockerfile#L7
combined with
https://github.com/docker/docker/blob/c21eb2a4209f6339fe59e4a97c3e6479aeb39f94/hack/make.sh#L109-L111
should be sufficient for that code to be enabled (and I actually used it on
1.9.1 successfully myself on Tuesday).

bobrik commented:
```
$ docker run -it --rm --log-driver=journald --name nope busybox ping google.com
```

```
$ docker logs nope
Error running logs job: configured logging reader does not support reading
```

```
$ docker version
Client:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.21
 Go version:   go1.4.2
 Git commit:   a34a1d5
 Built:        Fri Nov 20 12:59:02 UTC 2015
 OS/Arch:      linux/amd64
```

```
$ docker info
Containers: 263
Images: 1716
Server Version: 1.9.1
Storage Driver: overlay
 Backing Filesystem: xfs
Execution Driver: native-0.2
Logging Driver: json-file
Kernel Version: 4.1.3-wtf
Operating System: Debian GNU/Linux 8 (jessie)
CPUs: 32
Total Memory: 125.8 GiB
Name: 36com10
ID: CEGP:5ZLU:4ZGY:IQD7:RBMI:FB7A:JOB7:QP7Z:AXVM:N7Q5:Z5YH:YNLO
WARNING: No swap limit support
```

LK4D4 commented:
I think it's packaging problem. The binary must be linked with libsystemd.
Am I right, @tianon ?

tianon commented:
Interesting, I can reproduce on both 1.9.1 and 1.10.0-rc1 using the official `docker-engine` Debian packages (for `debian-jessie` specifically); both give something like:
```
Error running logs job: configured logging reader does not support reading
```

LK4D4 commented:
@tianon check ldd. Also, maybe we don't set `journald` tag.

nalind commented:
Hmm, I think the build containers (per contrib/builder/deb/generate.sh) might not have pkg-config in them.  I had erroneously assumed that they were pulled in by build-essential on Debian, and the various development tools groups.

If ``nm -CDu /usr/bin/docker | grep sd_journal_add_match`` turns up empty, then the logic's not getting compiled in.

tianon commented:
Yeah, I don't see `systemd` or `journal` anywhere in `ldd`, so I'll be digging into the build containers next.

tianon commented:
```console
$ nm -CDu /usr/bin/docker | grep sd_journal_add_match
$ 
```

Into the build system I go! :smile:

tianon commented:
Man, I could've sworn `pkg-config` was part of `build-essential` also, but it's not there!

dmp1ce commented:
I have Ubuntu 15.10 running with Docker version 1.10.0. I'm still getting `Error running logs job: configured logging reader does not support reading`. Should this have been fixed?

nalind commented:
Yes, it should have.  When I poke at the packages in the apt repository, I'm finding the expected references to the right unresolved symbols in the 1.10.1 packages for jessie, precise, trusty, and wheezy, but not in the packages for stretch or wily.

On stretch and wily, the expected headers are included in libsystemd-dev, which the build scripts install, but there is no libsystemd-journal.pc for pkg-config to find when hack/make.sh calls it, so the logic doesn't get built at build-time.  Given that those are newer versions, we might be better off changing pkg-config references to just look for libsystemd, and have make.sh look for a version of it that's at least 209, when systemd's news file says those were all merged together.

nalind commented:
Whoops, packages for precise and wheezy don't link with the right libraries - I was taking their packages apart wrong.  So only jessie and trusty are being built with the feature.

dmp1ce commented:
@nalind OK. Thanks. Is this already merged then for Wily? What about the future Xenial?

nalind commented:
@dmp1ce I was mistaken in thinking that we could fix it by changing the pkg-config dependency (we'd stop pulling things in on older releases, and there's no single option that's present across all of the desired build targets), so that was closed without merging.  So for the moment at least, it's not working in stretch or wily.

dmp1ce commented:
@nalind Should I create an issue for Wily?

nalind commented:
@dmp1ce Yeah, I think that's how we make sure it stays on the radar.
